airbyte-cdk/python/airbyte_cdk/sources/streams/http/availability_strategy.py

```diff
@@ -0,0 +1,133 @@
+            return True, None
+        except StopIteration:
+            logger.info(f"Successfully connected to stream {stream.name}, but got 0 records.")
+            return True, None
+        except HTTPError as error:
+            return self.handle_http_error(stream, logger, source, error)
+
+    def handle_http_error(
+        self, stream: Stream, logger: logging.Logger, source: Optional["Source"], error: HTTPError
+    ) -> Tuple[bool, Optional[str]]:
+        """
+        Override this method to define error handling for various `HTTPError`s
+        that are raised while attempting to check a stream's availability.
+
+        Checks whether an error's status_code is in a list of unavailable_error_codes,
+        and gets the associated reason for that error.
+
+        :param stream: stream
+        :param logger: source logger
+        :param source: optional (source)
+        :param error: HTTPError raised while checking stream's availability.
+        :return: A tuple of (boolean, str). If boolean is true, then the stream
+          is available, and no str is required. Otherwise, the stream is unavailable
+          for some reason and the str should describe what went wrong and how to
+          resolve the unavailability, if possible.
+        """
+        try:
+            status_code = error.response.status_code
+            reason = self.reasons_for_unavailable_status_codes(stream, logger, source, error)[status_code]
+            response_error_message = stream.parse_response_error_message(error.response)
+            if response_error_message:
+                reason += response_error_message
+            return False, reason
+        except KeyError:
+            # If the HTTPError is not in the dictionary of errors we know how to handle, don't except it
+            raise error
+
+    def reasons_for_unavailable_status_codes(
+        self, stream: Stream, logger: logging.Logger, source: Optional["Source"], error: HTTPError
+    ) -> Dict[int, str]:
+        """
+        Returns a dictionary of HTTP status codes that indicate stream
+        unavailability and reasons explaining why a given status code may
+        have occurred and how the user can resolve that error, if applicable.
+
+        :param stream: stream
+        :param logger: source logger
+        :param source: optional (source)
+        :return: A dictionary of (status code, reason) where the 'reason' explains
+        why 'status code' may have occurred and how the user can resolve that
+        error, if applicable.
+        """
+        forbidden_error_message = f"The endpoint to access stream '{stream.name}' returned 403: Forbidden. "
+        forbidden_error_message += "This is most likely due to insufficient permissions on the credentials in use. "
+        forbidden_error_message += self._visit_docs_message(logger, source)
+
+        reasons_for_codes: Dict[int, str] = {requests.codes.FORBIDDEN: forbidden_error_message}
```

Comment:
it might be worth parameterizing the class on this Dict. some APIs might return a 404 instead of a 403 if a set of credentials have insufficient permissions

Reply:
In that case, they'd also want to update the error message - in which case they can override the entire `reasons_for_unavailable_status_codes` method. Wdyt? Is it specifically declarative sources that makes us prefer a class variable over an overrideable method?

Reply:
I try to avoid redefining methods using inheritance because long inheritance chains can be tricky, but I see it's difficult to avoid because we need to Source object to produce the error message. Not something worth blocking on

Reply:
Haha, I do hear that - I had to write a separate doc just to figure out how all the inheritance was working for AbstractSource 😄